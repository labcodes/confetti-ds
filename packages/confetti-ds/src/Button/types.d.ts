export interface BaseButtonProps {
  /** This is the button label. */
  label: string;
  /** HTML type of the Button. */
  type?: "button" | "submit" | "reset";
  /** Sets the icon related to the button label. Default state: no icon. */
  hasIcon?: boolean;
  /* Enables the Button to render an Icon */
  icon?: IconTypes;
  /** Sets the button's height. Small = 32px, Normal = 40px, Large = 48px. */
  size?: "normal" | "small" | "large";
  /** Disables the Button. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled?: boolean;
  /** Disables the Button. Won't be read by screen readers. */
  disabled?: boolean;
  /** Action to be executed when the button is clicked. */
  onClick?: (event: SyntheticEvent) => any;
  /** Makes the button expand to its container's full width. */
  fullWidth?: boolean;
  /* Sets the order of which the elements will be focused on. Default value: 0. */
  tabIndex?: number;
}
