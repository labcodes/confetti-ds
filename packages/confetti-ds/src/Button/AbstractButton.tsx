import React, { SyntheticEvent } from "react";
import { isUndefined } from "lodash";

import Icon from "../Icon";
import { IconTypes } from "../constants";
import { BaseButtonProps } from "./types";

interface AbstractButtonProps extends BaseButtonProps {
  /** Button style variation. */
  kind?: "default" | "outline" | "text";
  /** Sets the icon related to the button label. Default state: no icon. */
  theme?:
    | "teal"
    | "purple"
    | "light"
    | "dark"
    | "warning"
    | "destructive"
    | "confirmation"
    | "warning-invert"
    | "destructive-invert"
    | "confirmation-invert";
}

export default function AbstractButton({
  label,
  type = "button" as const,
  kind,
  icon,
  theme = "teal",
  size = "small",
  ariaDisabled = false,
  disabled = false,
  onClick = (event) => {},
  fullWidth = false,
  tabIndex,
  hasIcon = false,
}: AbstractButtonProps) {
  const renderIcon = () =>
    icon ? (
      <Icon type={icon} color="white" className="lab-btn__icon" />
    ) : (
      <Icon type={"Add"} color="white" className="lab-btn__icon" />
    );
  /*
  Set icon color to be the icon color of default button.
  Icon color changes according to button type and exception on _buttons.scss
  */

  const handleOnClick = (event) => {
    if (!isUndefined(onClick)) {
      onClick(event);
    }
  };
  return (
    <button
      type={type}
      size={size}
      kind={kind}
      className={
        `lab-btn` +
        ` lab-btn--${kind} lab-btn--${size}` +
        `${theme ? ` lab-btn--${theme}` : ""}` +
        `${fullWidth ? ` lab-btn--block` : ""}`
      }
      onClick={!ariaDisabled ? handleOnClick : () => {}}
      disabled={(!ariaDisabled && disabled) || undefined}
      aria-disabled={ariaDisabled || undefined}
      {...(tabIndex ? { tabIndex } : undefined)}
    >
      {hasIcon ? renderIcon() : null}
      {label}
    </button>
  );
}
