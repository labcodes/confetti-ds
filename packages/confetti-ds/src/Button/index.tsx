export { default as Button, type ButtonProps } from "./Button";
export { default as OutlineButton, type OutlineButtonProps } from "./OutlineButton";
export { default as TextButton, type TextButtonProps } from "./TextButton";
