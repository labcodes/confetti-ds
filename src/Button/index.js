export { default as Button } from "./Button";
export { default as OutlineButton } from "./OutlineButton";
export { default as TextButton } from "./TextButton";
