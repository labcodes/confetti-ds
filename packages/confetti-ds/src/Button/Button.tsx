import React from "react";

import AbstractButton from "./AbstractButton";
import { BaseButtonProps } from "./types";
export interface ButtonProps extends BaseButtonProps {
  /** Button style variation. */
  variant?: "default" | "outline" | "text";
  /** Sets the icon related to the button label. Default state: no icon. */
  skin?:
    | ""
    | "light"
    | "dark"
    | "warning"
    | "destructive"
    | "warning-invert"
    | "destructive-invert"
    | "confirmation-invert";
}

export default function Button({
  text,
  type = "button" as const,
  icon,
  size = "normal",
  ariaDisabled = false,
  disabled = false,
  onClick = (event) => {},
  fullWidth = false,
  tabIndex,
  variant = "default",
  skin = "",
}: ButtonProps) {
  return (
    <AbstractButton
      variant="default"
      text={text}
      type={type}
      skin={skin}
      icon={icon}
      size={size}
      disabled={disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      fullWidth={fullWidth}
      tabIndex={tabIndex}
    />
  );
}
