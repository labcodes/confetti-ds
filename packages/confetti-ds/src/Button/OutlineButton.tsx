import React from "react";

import AbstractButton from "./AbstractButton";
import { BaseButtonProps } from "./types";

export interface OutlineButtonProps extends BaseButtonProps {
  skin?: "" | "light" | "dark";
}

export default function OutlineButton({
  text,
  type = "button",
  skin,
  icon,
  size = "normal",
  ariaDisabled = false,
  disabled = false,
  onClick = () => {},
  fullWidth = false,
  tabIndex,
}: OutlineButtonProps) {
  return (
    <AbstractButton
      variant="outline"
      text={text}
      type={type}
      skin={skin}
      icon={icon}
      size={size}
      disabled={disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      fullWidth={fullWidth}
      tabIndex={tabIndex}
    />
  );
}
