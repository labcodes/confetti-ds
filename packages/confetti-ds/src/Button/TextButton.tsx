import React from "react";

import AbstractButton from "./AbstractButton";
import { BaseButtonProps } from "./types";

export interface TextButtonProps extends BaseButtonProps {
  skin?: "" | "light" | "dark";
}

export default function TextButton({
  text,
  type = "button",
  skin,
  icon,
  size = "normal",
  ariaDisabled = false,
  disabled = false,
  onClick = () => {},
  fullWidth = false,
  tabIndex,
}: TextButtonProps) {
  return (
    <AbstractButton
      variant="text"
      text={text}
      type={type}
      skin={skin}
      icon={icon}
      size={size}
      disabled={disabled}
      ariaDisabled={ariaDisabled}
      onClick={onClick}
      fullWidth={fullWidth}
      tabIndex={tabIndex}
    />
  );
}
