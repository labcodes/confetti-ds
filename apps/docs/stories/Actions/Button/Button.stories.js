import React from "react";

import { ICON_TYPES } from "../../../../../packages/confetti-ds/src/constants";
import { Button as Component } from "../../../../../packages/confetti-ds/src/Button";

export default {
  title: "Actions/Button/Button",
  component: Component,
  argTypes: {
    icon: {
      control: { type: "select", options: ICON_TYPES },
    },
  },
};

export const Button = (args) => <Component {...args} />;
Button.args = {
  label: "Button",
  onClick: () => alert("Button was clicked"),
};
