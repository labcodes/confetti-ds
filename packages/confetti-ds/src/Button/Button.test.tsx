import React from "react";
import renderer from "react-test-renderer";
import { mount, shallow } from "enzyme";

import Button from "./Button";

describe("Button", () => {
  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<Button label="Test Default Button" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const renderedComponent = renderer
      .create(<Button label="Test Default Disabled Button 1" disabled />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <Button label="Test Default Disabled Button 2" disabled />
    );
    expect(mountedComponent.find("button").prop("disabled")).toEqual(true);
  });

  it("renders as expected when passing a warning skin", async () => {
    const renderedComponent = renderer
      .create(<Button label="Test Default Warning Button" theme="warning" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a small size", async () => {
    const renderedComponent = renderer
      .create(<Button label="Test Default Small Button" size="small" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a wallet icon", async () => {
    const renderedComponent = renderer
      .create(<Button label="Test Default Wallet Button" icon="wallet" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("calls props.onClick when clicked", async () => {
    const mockOnClick = jest.fn();
    const shallowButton = shallow(
      <Button label="Test Click on Default Button" onClick={mockOnClick} />
    );
    shallowButton.simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it("doesn't call onClick if ariaDisabled", async () => {
    const mockOnClick = jest.fn();
    const mountedComponent = mount(
      <Button ariaDisabled onClick={mockOnClick} label="Test Default Button" />
    );
    expect(mockOnClick.mock.calls.length).toEqual(0);
    mountedComponent.find("button").simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(0);
  });

  it("renders as expected if full width", async () => {
    const renderedComponent = renderer
      .create(<Button label="Test Default fullWidth Button" fullWidth />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with full width", async () => {
    const mountedComponent = mount(
      <Button label="Test fullWidth Button" fullWidth />
    );
    expect(mountedComponent.find(".lab-btn--block")).toHaveLength(1);
  });

  it("renders as expected when tabIndex is passed", async () => {
    const renderedComponent = renderer
      .create(
        <Button label="Test tabIndex" fullWidth tabIndex={1} />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });
});
