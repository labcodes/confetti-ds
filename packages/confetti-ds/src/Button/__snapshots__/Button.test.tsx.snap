// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders as expected if full width 1`] = `
<button
  className="lab-btn lab-btn--default lab-btn--normal lab-btn--block"
  onClick={[Function]}
  type="button"
>
  
  Test Default fullWidth Button
</button>
`;

exports[`Button renders as expected when passing a small size 1`] = `
<button
  className="lab-btn lab-btn--default lab-btn--small"
  onClick={[Function]}
  type="button"
>
  
  Test Default Small Button
</button>
`;

exports[`Button renders as expected when passing a wallet icon 1`] = `
<button
  className="lab-btn lab-btn--default lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  <span
    className="lab-icon lab-icon--wallet lab-icon--white lab-icon--petit lab-btn__icon"
  />
  Test Default Wallet Button
</button>
`;

exports[`Button renders as expected when passing a warning skin 1`] = `
<button
  className="lab-btn lab-btn--default lab-btn--normal lab-btn--warning"
  onClick={[Function]}
  type="button"
>
  
  Test Default Warning Button
</button>
`;

exports[`Button renders as expected when passing disabled as true 1`] = `
<button
  className="lab-btn lab-btn--default lab-btn--normal"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  
  Test Default Disabled Button 1
</button>
`;

exports[`Button renders as expected when tabIndex is passed 1`] = `
<button
  className="lab-btn lab-btn--default lab-btn--normal lab-btn--block"
  onClick={[Function]}
  tabIndex={1}
  type="button"
>
  
  Test tabIndex
</button>
`;

exports[`Button renders with base props 1`] = `
<button
  className="lab-btn lab-btn--default lab-btn--normal"
  onClick={[Function]}
  type="button"
>
  
  Test Default Button
</button>
`;
